Allow for Persons and Organisations to be Correspondents

BalduinLandolt opened the pull request:
This PR adds an abstract class `Correspondent`, and makes both `Person` and `Organisation` children of `Correspondent.`

Then, `get_author()`, and `get_recipient()` respectively, can return a `Correspondent`. Thus, not only persons, but also organisations can be authors or recipients.

Fixes #28

BalduinLandolt committed:
Feature (data structure): Add abstract class Correspondent; make Person and Organisation children of Correspondent

BalduinLandolt committed:
Feature (correspondent): look up organisations for authors and recipients

BalduinLandolt committed:
Feature (Test): test organisations in recipient

BalduinLandolt committed:
Feature (correspondent): improve type identification of correspondent

BalduinLandolt commented:
@tobiasschweizer several things:

- I don't have an example of an organisation being author, so I could test that. Need I bother Basil, or do we trust i works, if it works for recipients?
- At the moment it's a bit inconsistent: we have a method `get_recipient_organisation()`, but neither `get_recipient_person()` nor `get_author_organisation()` and `get_author_person()`. So either we should have a method for all of these cases, or none.
- depending on wether or not we keep these methods, everything or nothing in `get_author()` and `get_recipient()` should be done like this: https://github.com/dhlab-basel/alephmarcreader/blob/5d25eac11ffaebed50a4cb6bcdbf785a3bcfa5a7/alephmarcreader/abstractalephmarcreader.py#L320

tobiasschweizer commented:
Nice! I will try to have a look at it next week!

tobiasschweizer commented:
> At the moment it's a bit inconsistent: we have a method get_recipient_organisation(), but neither get_recipient_person() nor get_author_organisation() and get_author_person(). So either we should have a method for all of these cases, or none.

I think there should just be a method `get_recipient` that gets both persons and organisations, like you did for authors.

Are there mentioned organisations?

tobiasschweizer commented on alephmarcreader/abstractalephmarcreader.py:
```diff
@@ -7,39 +7,67 @@
+
+        """
+
+        CORRESPONDENT_TYPE_PERSON = 'person'
```
This should have a capital "p" like the class name.

BalduinLandolt replied:
done

tobiasschweizer commented on alephmarcreader/abstractalephmarcreader.py:
```diff
@@ -7,39 +7,67 @@
+        """
+
+        CORRESPONDENT_TYPE_PERSON = 'person'
+        CORRESPONDENT_TYPE_ORGANISATION = 'organisation'
```
same as above

BalduinLandolt replied:
done

tobiasschweizer commented on alephmarcreader/tests/test_Marc21Reader.py:
```diff
@@ -28,6 +28,22 @@ def test_recipient(self):
+
+        self.assertEqual(len(recipient), 2)
+
+        self.assertEqual(recipient[0].name, u'Universit\xe4t Basel.')
```
please add a check for the type (person or organisation), both for the string constant (`get_type`) an the actual instance (Person, Organisation)

tobiasschweizer replied:
actually the same would have to be done for the already existing tests for persons

BalduinLandolt replied:
done

BalduinLandolt committed:
Fix (Correspondent): change capitalization of correspondent type, so that it's the same as class name

BalduinLandolt committed:
Feature (Test): Add test for class name and type for Persons and Organistaions

BalduinLandolt committed:
Feature (Recipient): Remove `get_recipient_organisation()` as it is no longer needed.
(`get_recipient()` returns Correspondent which can be both Person or Organisation)

BalduinLandolt committed:
Fix (Recipient): Remove uses of `get_recipient_organisation()` that no longer exists

BalduinLandolt commented:
> 
> 
> > At the moment it's a bit inconsistent: we have a method get_recipient_organisation(), but neither get_recipient_person() nor get_author_organisation() and get_author_person(). So either we should have a method for all of these cases, or none.
> 
> I think there should just be a method `get_recipient` that gets both persons and organisations, like you did for authors.
> 
> Are there mentioned organisations?

Done.

We now have `get_author` and `get_recipient` which both return `Correspondent` which can be `Person` or `Organisation`.  
For mentioneds on the other hand, we now have `get_mentioned_person` and `get_mentioned_organisation` which are specific to a type each.

This is not exactly consistent, but in terms of usability, it seems like what is most practical. In that sense, I'm not sure, I'd want to get rid of these methods.  
What we could do, though, is to add a conveinience method `get_mentioned_correspondent` that simply returns the combination of `get_mentioned_person` and `get_mentioned_organisation`.

Or how do you feel about this, @tobiasschweizer ?

tobiasschweizer commented on alephmarcreader/abstractalephmarcreader.py:
```diff
@@ -7,39 +7,67 @@
+            Gives the type of Correspondent.
+            :return str: Type of correspondent, either 'person' or 'organisation'.
+            """
+            pass
```
Is the `pass` necessary? The method is abstract and does not need an implementation.

BalduinLandolt replied:
I think it needs that... I may have googled "python abstract class" and have found something like [this](https://www.python-course.eu/python3_abstract_classes.php), where they use `pass`.

But I can't say for sure, that it's necessary. What I can say for sure, though, is that it works as it is now.

tobiasschweizer replied:
I think you are right, we used it for other abstract methods too. I think I simply coded too much TypeScript ...

tobiasschweizer commented on alephmarcreader/abstractalephmarcreader.py:
```diff
@@ -267,13 +295,20 @@ def get_author(self):
             if "aut" in person.roles:
                 author.append(person)
 
```
Shouldn't there be a check for the filed 110 ("110 = Haupteintragung unter einem Körperschaftsnamen") as well?

BalduinLandolt replied:
Technically, yes. But in BIBB, field 110 has never been used. In the instances, where the author is an institution, that institution is named in field 710 with the role 'aut' (see e.g. [000058439](https://aleph.unibas.ch/F/4BNNT8PNYCPGK8CICMQA9MPE6194BYLHD5VJEIGQN271FFG6IF-08311?func=full-set-set&set_number=672222&set_entry=000001&format=001), field `710 2`).  
So there's really no point in checking an unused field.

tobiasschweizer replied:
Ok, got it!

tobiasschweizer reviewed (commented):
This looks good now!

Could you make version 1.1.0 in https://github.com/dhlab-basel/alephmarcreader/blob/364839fcabd0210f8265347ecbeee947b0e472c5/setup.py#L8? 

We should also have a change log file. Could you explain in there that the return type changed from `Person` to `Organisation` and that `get_recipient_organisation` does not exist any more?

BalduinLandolt commented:
> 
> 
> This looks good now!
> 
> Could you make version 1.1.0 in
> 
> https://github.com/dhlab-basel/alephmarcreader/blob/364839fcabd0210f8265347ecbeee947b0e472c5/setup.py#L8
> ?

Will do.

> 
> We should also have a change log file. Could you explain in there that the return type changed from `Person` to `Organisation` and that `get_recipient_organisation` does not exist any more?

Aye. Do you want this as a separate file? Or could we simply make a GitHub release once it's merged to the master branch, and put the change log there?

BalduinLandolt committed:
Feature: Set version number to 1.1.0

BalduinLandolt commented:
> Or could we simply make a GitHub release once it's merged to the master branch, and put the change log there?

in that case, we should also have a release 1.0.0 from the current master (before we merge this)

tobiasschweizer commented:
> Do you want this as a separate file? Or could we simply make a GitHub release once it's merged to the master branch, and put the change log there?

I think it's good to a have a file like `release-notes.md`. But it doesn't hurt to describe the gitHub release as well.

> in that case, we should also have a release 1.0.0 from the current master (before we merge this)

good thinking!

BalduinLandolt commented:
> 
> 
> > Do you want this as a separate file? Or could we simply make a GitHub release once it's merged to the master branch, and put the change log there?
> 
> I think it's good to a have a file like `release-notes.md`. But it doesn't hurt to describe the gitHub release as well.
> 
> > in that case, we should also have a release 1.0.0 from the current master (before we merge this)
> 
> good thinking!

on it

tobiasschweizer reviewed (approved):
Once #30 is merged and this branch is up to date with `develop`, please update the release notes on this branch before merging.

BalduinLandolt committed:
Pull Release Notes to WIP-Branch

* Feature: Add release notes for v1.0.0

* Refactor (sample usage): Move example usage to README.md

BalduinLandolt committed:
Feature (release notes): Add release notes for v1.1.0

BalduinLandolt committed:
Merge branch 'master' into wip/correspondent

BalduinLandolt commented:
@tobiasschweizer sorry for the slight chaos I caused... (by accident, I first pushed the updated release notes to the master branch, which I had to revert.)

If you approve of the release notes, I think it would be ready for merging and releasing as v1.1.0.

(Also, see the 1.0.0 release. I hope that's ok like this.)

tobiasschweizer commented:
no problem, I think you are good to go now
